compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -1430,23 +1431,34 @@ public BType transform(BLangFiniteTypeNode finiteTypeNode, AnalyzerData data) {
 
     @Override
     public BType transform(BLangTupleTypeNode tupleTypeNode, AnalyzerData data) {
-        List<BType> memberTypes = new ArrayList<>();
-        for (BLangType memTypeNode : tupleTypeNode.memberTypeNodes) {
-            BType type = resolveTypeNode(memTypeNode, data, data.env);
-
-            // If at least one member is undefined, return noType as the type.
-            if (type == symTable.noType) {
-                return symTable.noType;
-            }
-
-            memberTypes.add(type);
-        }
-
-        BTypeSymbol tupleTypeSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(EnumSet.of(Flag.PUBLIC)),
-                Names.EMPTY, data.env.enclPkg.symbol.pkgID, null,
+        List<BLangSimpleVariable> members = new ArrayList<>(tupleTypeNode.members.size());
+        BTypeSymbol tupleTypeSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE,
+                Flags.asMask(EnumSet.of(Flag.PUBLIC)), Names.EMPTY, data.env.enclPkg.symbol.pkgID, null,
                 data.env.scope.owner, tupleTypeNode.pos, SOURCE);
-
-        BTupleType tupleType = new BTupleType(tupleTypeSymbol, memberTypes);
+        SymbolEnv tupleEnv = SymbolEnv.createTypeEnv(tupleTypeNode, new Scope(tupleTypeSymbol), data.env);
+        boolean hasUndefinedMember = false;
+        for (BLangSimpleVariable member : tupleTypeNode.members) {
+            BType bType = member.getBType();
+            if (bType == null) {
+                symbolEnter.defineNode(member, tupleEnv);
+            } else if (bType == symTable.noType) {
+                member.setBType(null);
+                symbolEnter.defineNode(member, tupleEnv);
+            }
+            if (member.getBType() == symTable.noType) {
+                hasUndefinedMember = true;
+            }
+            members.add(member);
+        }
+        // If at least one member is undefined, return noType as the type.
+        if (hasUndefinedMember) {
+            return symTable.noType;
+        }
```

Comment:
```ballerina
type Foo [int, int, Bar];
type Bar int;
```
With this approach, if we consider the above example, we first resolve the first two members and identify that the "Bar" type is not resolved. Then, we return it and identify that the "Foo" type is not resolved. However, the actual not resolved type is "Bar." To correct this issue for the "function type," we resolve the member types after resolving all the type definitions.

https://github.com/ballerina-platform/ballerina-lang/blob/f715e82171cfac4a4d8010e27ff29177a66d532f/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java#L440

Reply:
`Bar` not being resolved means `Foo` isn't resolved either. IMO, that is OK. IIRC, that's what's done for records too.

Reply:
Yes, it is fine. However, we need to resolve member types of type definitions after resolving all the type definitions. If not, we will need to perform additional work. As previously mentioned, when we identify that 'Foo' is not defined, we will need to attempt to define member types again. However, during the previous iteration, the types of some members may have already been defined. We need to check that.

Reply:
These should ideally be resolved with the symbol enter revamp, since we decided to define the "shell" of the type first and then resolve members.